### ISSUE TITLE
Nodes stuck in decommissioning state

### ISSUE THREAD

**Issue by aaitken-apixio** [cockroachdb/cockroach#35156]
**Summary**
We attempted to decommission a number of nodes a few weeks back and unfortunately didnt do it all in one command. I ran a decommission on a single node followed by additional decommission commands on other nodes before the first had complete. We are now stuck in a state where we have a 12 node cluster with 6 of those nodes decommissioning but wont finish.
Currently our cluster overview shows 6 live nodes and 6 suspect nodes as well as 5 under-replicated ranges. On the problem ranges report it shows that we have 5 ranges with no raft leader and 2 ranges which are under-replicated (slow). 
The 6 suspect nodes have these ranges with either no raft leader or under-replicated ranges. We cannot attempt to recommission the nodes as they are stuck in the decommissioning state. If we shut down any of the suspect nodes then some of the problem ranges now become unavailable meaning we cant query any data associated with them. So we are stuck atm with the additional 6 suspect nodes which we cannot get rid of and cannot fix the issue with the problem ranges.

**Steps to reproduce**
Not sure exactly how to reproduce but guessing it has to do with us running the decommission tool multiple times before waiting for some of them to fully complete. 

**Expected Result**
CockroachDB should know which other nodes are being decommissioned and rebalance the data accordingly. Would also expect that CockroachDB could force a new raft leader for these problem ranges. Currently all data is queryable but not if we lose any 1 of the currently suspect nodes

**Actual Result**
Seems like CockroachDB got into problems trying to run multiple decommission tasks simultaneously. It also cannot recover from having no raft leader for ranges on nodes which are attempting to decommission. 

**Log files/version**
We are running version 2.1.3. 
Below are an example of logs from one of our suspect nodes:

W190222 18:20:14.900777 291915892 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:14.943744 291915808 storage/store.go:3812  [n3,s4,r6630/2:/Table/71/9/"ANNOTATE"/41…] raft error: node 8 claims to not contain store 9 for replica (n8,s9):1: store 9 was not found
W190222 18:20:14.943760 291915806 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:15.100743 291915863 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:15.100759 291915861 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:15.300812 291915866 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:15.300833 291915864 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
I190222 18:20:15.434721 427 gossip/gossip.go:557  [n3] gossip status (ok, 13 nodes)
gossip client (0/3 cur/max conns)
gossip server (2/3 cur/max conns, infos 152297623/141231374 sent/received, bytes 60411527407B/47207748704B sent/received)
  5: 10.0.32.23:26257 (1358h7m55s)
  8: 10.0.51.34:26257 (1358h7m55s)
W190222 18:20:15.500830 291915894 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:15.500849 291915877 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:15.595708 291915837 storage/store.go:3812  [n3,s4,r5608/11:/Table/114/1/405{8603…-9564…}] raft error: node 8 claims to not contain store 9 for replica (n8,s9):6: store 9 was not found
W190222 18:20:15.595725 291915835 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:15.700884 291915881 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:15.700900 291915879 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:15.900864 291915841 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:15.900883 291915839 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:16.100790 291915886 storage/store.go:3812  [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found
W190222 18:20:16.100809 291915884 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:
W190222 18:20:16.107207 291915908 storage/store.go:3812  [n3,s4,r5608/11:/Table/114/1/405{8603…-9564…}] raft error: node 8 claims to not contain store 9 for replica (n8,s9):6: store 9 was not found
W190222 18:20:16.107221 291915842 storage/raft_transport.go:584  [n3] while processing outgoing Raft queue to node 8: store 9 was not found:

**Comment by tbg** [comment]
Sorry for the delay, @aaitken-apixio. @tim-o could you take a look and qualify further?

What you were doing should work. Are you by any chance reusing any decommissioned nodes as new nodes? Or asked differently, was your node initially a 12 node cluster and you tried to decommission six of its nodes? Or did the nodes in your cluster perhaps "cycle" their IPs (i.e. the host:port for n8 is now the host:port for n9 or similar, this happens frequently in k8s deploys etc)

The reason I'm asking is because this message

> W190222 18:20:15.900864 291915841 storage/store.go:3812 [n3,s4] raft error: node 8 claims to not contain store 9 for replica (n8,s9):?: store 9 was not found

is usually caused by (in this example) a node trying to connect to n8 but really accidentally reaching another node in the cluster, which then won't have the data the requester is looking for.

We have fixes for this problem in the pipeline, but they're not in 2.1.3. I believe they are in v2.1.5, though I'll want to confirm that.

@knz I think the relevant PR is #34198. Could you verify that the situation here matches? It doesn't seem that we accidentally removed any node descriptors from Gossip here. Rather the mapping nodeID -> IP seems to be off.

**Comment by knz** [comment]
Yes it matches the symptoms of #34198 which I think will solve the unavailability problem. Although even with that PR the same log messages might still appear even after the cluster becomes functional again. To make the log files fully disappear will need #34197 too.
